### PR TITLE
Close queue and join_thread on handler close

### PR DESCRIPTION
**Pull request by al4**
Hi there!

TL;DR: This PR closes the queue and joins to ensure it has been flushed on application shutdown. I *think* it is safe. The tests pass, it works for me, and it seems "correct", but I'm not 100% sure it will work for everyone. I'd be interested to know your thoughts.

First a bit of background - I'm using multiprocessing-logging in an application that can be invoked as a running service or a command line application.

In the case of the service, everything works really well. However in the command line application, the last couple of logging calls often aren't emitted before the application exits. This can result in a random exception, but usually the messages just disappear. I suspect the same thing is happening in the service case, but being masked by slower termination, and the fact that it's not as noticeable if a random log line is missed as opposed to the final output of your terminal app.

The fix, in my case, is the contents of this pull request. When close() is invoked on the handler, it closes the queue and then joins it, waiting for it to be flushed (https://docs.python.org/3.5/library/multiprocessing.html#multiprocessing.Queue.close). The question, is whether this is safe for everyone's use cases.

The immediate scenario that comes to mine is if thread A calls Queue.close(), and thread B subsequently calls put(). According to the docs, Queue.close() indicates that "no more data will be put on this queue by the current process". Thus, it seems fair to assume it will not affect other processes.

Also, close() is supposed to only be called by logging.shutdown(), which states "This should be called at application exit and no further use of the logging system should be made after this call."
So it would be incorrect for a put to be made to the queue after close().

### CODE REVIEW

**Comment by philipaconrad** [comment]
I've run into the problem @al4 describes (final log messages not getting flushed) in my own applications using `multiprocessing-logging`.

### Use case

A long-running service that forks off child processes to do various jobs.

### Why the flush bug affects me

I like to see that the various job processes all shut down as expected.
When the output is not present in the logs, I have nothing to go on.

### Problems with the Pull Request

I sometimes still don't see the final log messages from the main process in my output logs.

When the last few lines from that process are present, an exception is thrown (output shown below):

```
[INFO] [Job-1] [2016-12-27 03:37:30,505] Terminating normally.
[INFO] [Job-2 ] [2016-12-27 03:37:30,506] Terminating normally.
[INFO] [MainProcess] [2016-12-27 03:37:30,518] Shutting down.
Traceback (most recent call last):
  File "/<REDACTED>/venv/local/lib/python2.7/site-packages/multiprocessing_logging.py", line 55, in receive
```

System Details:

 - **OS**: Ubuntu 16.04
 - **Python Version**: 2.7.10

**Comment by al4** [comment]
@philipaconrad it still seems to happen on our end as well, I'll update here if I find a fix (it might be outside of this library).

For now, consider this PR incomplete.

**Comment by jruere** [comment]
It would be great to be able to reproduce. Could you provide a minimal case?

**Comment by al4** [comment]
Running the code as this PR currently stands, I have observed the following exception (this does not happen consistently):
```
Exception in thread mp-handler-2 (most likely raised during interpreter shutdown):
Traceback (most recent call last):
  File "/usr/lib/python2.7/threading.py", line 810, in __bootstrap_inner
  File "/usr/lib/python2.7/threading.py", line 763, in run
  File "/opt/venvs/##redacted##/local/lib/python2.7/site-packages/multiprocessing_logging.py", line 62, in receive
<type 'exceptions.AttributeError'>: 'NoneType' object has no attribute 'print_exc'```
```
print_exc is actually a function of the traceback module, the code that bailed here is:
```
 def receive(self):
     while True:
         try:
             record = self.queue.get()
             self.sub_handler.emit(record)
         except (KeyboardInterrupt, SystemExit):
             raise
         except EOFError:
             break
         except:
             traceback.print_exc(file=sys.stderr)
```
It appears that we tried to retrieve a message after the traceback module had been torn down. It is possible that the exception which _would_ have been printed would reveal that the sub_handler (or the queue) have been torn down, hence the exception. 

Anyway I still haven't ruled out a problem with my own code yet, will keep digging.

**Comment by jruere** [comment]
@al4 I don't mean about the problem @philipaconrad reports but about the problem the PR addresses. If we have code that shows the problem, it will be easier to discuss.
I'll try to work on this this weekend.

Bear in mind that we are not even trying to keep all messages. For example, on `send()`, we [discard messages if the queue is full](https://docs.python.org/3.5/library/multiprocessing.html#multiprocessing.Queue.get_nowait).

It's a good thing to delay termination if we are flushing messages though.

**Comment by jruere** [comment]
On the code without the PR, if the forked process is daemon, I can reproduce this problem.

```python
import logging
import multiprocessing
import sys

import multiprocessing_logging


logging.basicConfig(level=logging.DEBUG)

multiprocessing_logging.install_mp_handler()


def task(id):
    logging.info('Task started: %s', id)
    import time
    time.sleep(1)
    logging.info('Task complete: %s', id)


logging.info('Starting tasks...')

p1 = multiprocessing.Process(target=task, args=(1,), daemon=True)
p1.start()

p2 = multiprocessing.Process(target=task, args=(2,))
p2.start()

logging.info('Started tasks.')

```

Non daemon processes don't lose records.

Are you guys using daemon processes?

**Comment by mikeengland** [comment]
Is there any further progress on this?

**Comment by jruere** [comment]
@mikeengland are you using daemon processes?

**Comment by mikeengland** [comment]
@jruere Yeah, they are daemons.

**Comment by mikeengland** [comment]
@jruere Is there any update on a fix for this?

**Comment by jruere** [comment]
I'm not convinced this fix is good for this case. It would block the main
thread until the children finish so it makes no sense to use daemon threads.

I haven't had time to play with it.

**Comment by al4** [comment]
Yes I don't have any confidence that this is the right solution either, and never did get around to digging deeper into it.

Also, I have since changed companies and no longer have access to the code which triggered this problem. I'll close this then, as the commit shouldn't be merged as it stands.
